code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/XDMUtils.kt

```diff
@@ -0,0 +1,102 @@
+  you may not use this file except in compliance with the License. You may obtain a copy
+  of the License at http://www.apache.org/licenses/LICENSE-2.0
+  Unless required by applicable law or agreed to in writing, software distributed under
+  the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+  OF ANY KIND, either express or implied. See the License for the specific language
+  governing permissions and limitations under the License.
+*/
+
+package com.adobe.marketing.mobile.optimize
+
+import com.adobe.marketing.mobile.Event
+import com.adobe.marketing.mobile.MobileCore
+import com.adobe.marketing.mobile.services.Log
+
+object XDMUtils {
+
+    private const val SELF_TAG: String = "XDMUtils"
+
+    /**
+     * Dispatches an event to track propositions with type {@value
+     * * OptimizeConstants.EventType#OPTIMIZE} and source {@value
+     * * OptimizeConstants.EventSource#REQUEST_CONTENT}.
+     * No event is dispatched if the provided `xdm` is null or empty.
+     * @param xdm `Map<String, Object>` containing the XDM data for the proposition
+     * interactions.
+     */
+    @JvmStatic
+    fun trackWithData(xdm: Map<String, Any>?) = xdm?.takeIf { it.isNotEmpty() }?.let {
+        val eventData: MutableMap<String, Any> = hashMapOf(
+            OptimizeConstants.EventDataKeys.REQUEST_TYPE to OptimizeConstants.EventDataValues.REQUEST_TYPE_TRACK,
+            OptimizeConstants.EventDataKeys.PROPOSITION_INTERACTIONS to it
+        )
+        val edgeEvent = Event.Builder(
+            OptimizeConstants.EventNames.TRACK_PROPOSITIONS_REQUEST,
+            OptimizeConstants.EventType.OPTIMIZE,
+            OptimizeConstants.EventSource.REQUEST_CONTENT
+        ).setEventData(eventData).build()
+
+        MobileCore.dispatchEvent(edgeEvent)
+    } ?: Log.debug(
+        OptimizeConstants.LOG_TAG,
+        SELF_TAG,
+        "Failed to dispatch track propositions request event, input xdm is null or empty."
+    )
+
+    /**
+     * Generates a map containing XDM formatted data for `Experience Event - OptimizeProposition
+     * Interactions` field group from this `OptimizeProposition` offer and given
+     * `experienceEventType`.
+     * The method returns null if the proposition reference within the offer is released and no
+     * longer valid.
+     * @param experienceEventType [String] containing the event type for the Experience Event
+     * @return `Map<String></String>, Object>` containing the XDM data for the proposition interaction.
+     */
+
+    @JvmStatic
+    fun generateInteractionXdm(
+        experienceEventType: String,
+        pendingInteraction: InteractionPropositionType
+    ): Map<String, Any> {
+
+        val finalPendingPropositions = when (pendingInteraction) {
+            is InteractionPropositionType.MultiplePropositions -> pendingInteraction.propositions
+            is InteractionPropositionType.SingleProposition -> listOf(pendingInteraction.proposition)
```

Comment:
Do we need this new `InteractionPropositionType`? Can we not accept a list of propositions instead? We're converting the single proposition also to a list in the end anyway

Reply:
Nevermind, just saw line 74